File: 22-walls-and-gates/20221207/walls-and-gates.js
"use strict";

const GATE = 0;
const DIRECTIONS = [
    [-1, 0],
    [0, 1],
    [1, 0],
    [0, -1],
];

function wallsAndGates(rooms) {
    for (let i = 0; i < rooms.length; ++i) {
        for (let j = 0; j < rooms[0].length; ++j) {
            if (rooms[i][j] === GATE) {
                dfs(rooms, i, j, 0);
            }
        }
    }

    return rooms;
}

function dfs(rooms, row, col, step) {
    if (!inBounds(row, col, rooms) || (step && step >= rooms[row][col])) {
        return;
    }

    rooms[row][col] = step;

    for (const DIR of DIRECTIONS) {
        const [newRow, newCol] = [row + DIR[0], col + DIR[1]];
        dfs(rooms, newRow, newCol, step + 1);
    }
}

function inBounds(row, col, grid) {
    return row >= 0 && row < grid.length && col >= 0 && col < grid[0].length;
}

module.exports = wallsAndGates;

File: 22-walls-and-gates/20221207/walls-and-gates.py
from typing import List


class Solution:
    GATE = 0
    DIRECTIONS = [[-1,0], [0,1], [1,0], [0,-1]]

    def wallsAndGates(self, rooms: List[List[int]]) -> None:
        for i in range(len(rooms)):
            for j in range(len(rooms[0])):
                if rooms[i][j] == self.GATE:
                    self.dfs(rooms=rooms, row=i, col=j)
        return None

    def dfs(self, rooms: List[List[int]], row: int, col: int, step: int=0) -> None:
        if not self.in_bounds(rooms, row, col) or (step and step >= rooms[row][col]):
            return None

        rooms[row][col] = step

        for [row_mod, col_mod] in self.DIRECTIONS:
            self.dfs(rooms, row + row_mod, col + col_mod, step + 1)

        return None

    def in_bounds(self, rooms: List[List[int]], row: int, col: int) -> bool:
        return 0 <= row < len(rooms) and 0 <= col < len(rooms[0])
